File: dungeon_base/src/battle/ref.erl
-module(ref).

-compile({no_auto_import,[get/1]}).
-export([get/1, get/3, set/2, set/4, val/1, val/3, who/3]).

who(off, O, _D) -> O;
who(def, _O, D) -> D.

get({attr, Attr, Whose}) ->

    case Whose of
        #{attr  :=#{Attr:=Val}} -> Val;
        #{state :=#{Attr:=Val}} -> Val
    end.


set({attr, Attr, Whose}=Ref, NewVal) when is_number(NewVal) ->

    case Whose of
        #{state:=#{Attr:=_}=AttrSet, attr:=AttrTypeset} ->
            Whose#{state:=AttrSet#{
                            Attr:={single, round(NewVal)}
                           },
                   attr:=AttrTypeset#{
                           diff:={single, round(NewVal-val(Ref))}
                          }
                  };
        #{attr :=#{Attr:=_}=AttrSet} ->
            Whose#{attr:=AttrSet#{
                           Attr:={single, round(NewVal)},
                           diff:={single, round(NewVal-val(Ref))}
                          }
                  }
    end;

set({attr, Attr, Whose}, Val) ->

    case Whose of
        #{attr :=#{Attr:=_} = AttrSet} ->
            Whose#{attr  :=AttrSet#{Attr:=Val}};
        #{state :=#{Attr:=_} = AttrSet} ->
            Whose#{state :=AttrSet#{Attr:=Val}}
    end.



val({range, [Low, High]}) ->
    round(Low + rand:uniform() * (High - Low));

val({single, SingleValue}) -> SingleValue;

val({attr, Attr, Whose} = _Ref) ->

    case Whose of
        #{attr  :=#{Attr:=Val}} -> val(Val);
        #{state :=#{Attr:=Val}} -> val(Val)
    end.


get({attr, Attr, Whose}, O, D) ->
    get({attr, Attr, who(Whose, O, D)}).


val({attr, Attr, Whose}, O, D) ->
    val({attr, Attr, who(Whose, O, D)});
val(Other, _O, _D) ->
    val(Other).

set({attr, Attr, Whose}, O, D, Val) ->
    set({attr, Attr, who(Whose, O, D)}, Val).

File: dungeon_elixir/lib/battle_logic/ref.ex
defmodule Ref do
	
	# 返回角色
	def get_role(:off, o, _d), do: o
	def get_role(:def, _o, d), do: d

	def get({:attr, type, attr, player}) do
		%{^type => %{^attr => value}} = player
		val(value)
	end

	def set({:attr, type, attr, player}=ref, value) when is_number(value) do
		%{^type => type_set} = player
		%{player | type => %{ type_set | attr => {:single, round(value)}, :diff => {:single, round(value-val(ref))} }}
	end

	def set({:attr, type, attr, player}, value) do
		%{^type => type_set} = player
		%{player | type=> %{ type_set | attr=> value }}
	end


	def val({:range, low, high}), do: round(low + :rand.uniform() * (high - low))
	def val({:single, single_value}), do: single_value

	def val({:attr, type, attr, player}) do
		%{^type => %{^attr => value}} = player
		val(value)
	end

	def get({:attr, type, attr, role}, o, d) do
		get({:attr, type, attr, get_role(role, o, d)})
	end


	def val({:attr, type, attr, role}, o, d), do: val({:attr, type, attr, get_role(role, o, d)})
	def val(other, _o, _d), do: val(other)

	def set({:attr, type, attr, role}, o, d, value) do
		set({:attr, type, attr, get_role(role, o, d)}, value)
	end

end